src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleaner.py

```diff
@@ -571,8 +582,39 @@ def getMSOutputTransferInfo(self, wflow):
             msg = "General exception while fetching TransferInfo from MSOutput for %s. "
             msg += "Error: %s"
             self.logger.exception(msg, wflow['RequestName'], str(ex))
+
+        # Set Transfer status - information fetched from MSOutput only
         if transferInfo is not None and transferInfo['TransferStatus'] == 'done':
             wflow['TransferDone'] = True
+
+        # Set Tape rules status - information fetched from Rucio (tape rule ids from MSOutput)
+        if transferInfo is not None and transferInfo['OutputMap']:
+            tapeRulesStatusList = []
+            # For setting 'TransferTape' = True we require either no tape rules for the
+            # workflow have been created or all existing tape rules to be in status 'OK',
+            # so every empty TapeRuleID we consider as completed.
+            for ruleId in [mapRecord['TapeRuleID'] for mapRecord in transferInfo['OutputMap']]:
+                if not ruleId.strip():
```

Comment:
Updating my own comment, it's actually not O(n^2). However, I still think it's a valid suggestion above.

Reply:
Thinking again, just disregard this comment, not worth it! Sorry for the noise.

Reply:
Thanks @amaltaro 
>  to avoid O(n^2) complexity, how about changing the code to something like:

I can not see what you mean by this, because if you refer to the way how I build the list for the loop, I am quite sure it is `O(n)`. The list comprehension happens in place and is executed only once at the beginning of the `for` cycle, and since then it is kept in memory as a pure list for the consequent iterations. This is at least my understanding, and just to confirm it I did step through that cycle few times with pdb and it did build the list in advance. But I agree with you, for the sake of shortening the line and improving readability your suggestion has quite of a value. And also for not keeping in memory a list (even a short one) of something to which we already have a chain of references in memory. I am about to use it.

The other two comments you made here I really liked and I did manage to reduce the depth of `if` conditions here. Thanks  for that too.

Reply:
Done